driver/amc_pci_core.c

```diff
@@ -160,15 +204,45 @@ static int create_device_nodes(
 
     cdev_init(&amc_priv->cdev, &base_fops);
     amc_priv->cdev.owner = THIS_MODULE;
-    int rc = cdev_add(&amc_priv->cdev, MKDEV(major, minor), MINORS_PER_BOARD);
+    int rc = cdev_add(&amc_priv->cdev, MKDEV(major, minor),
+        amc_priv->prom->nentries);
     TEST_RC(rc, no_cdev, "Unable to add device");
 
-    for (int i = 0; i < MINORS_PER_BOARD; i ++)
-        device_create(
-            device_class, &pdev->dev, MKDEV(major, minor + i), NULL,
-            "%s.%d.%s", DEVICE_NAME, amc_priv->board, fops_info[i].name);
+    size_t i = 0;
+    char *device_name = NULL;
+    struct prom_entry *pentry;
+    prom_for_each_entry(pentry, amc_priv->prom)
+    {
+        // it is assumed that the device entry is the first found
+        if (pentry->tag == PROM_DEVICE_TAG)
+        {
+            struct prom_device_entry *device_entry =
+                (struct prom_device_entry *) pentry;
+            device_name = device_entry->name;
+            device_create(
+                device_class, &pdev->dev, MKDEV(major, minor + i), NULL,
+                "%s.%d.reg", device_name, amc_priv->board);
+        }
+        else if (pentry->tag == PROM_DMA_TAG)
+        {
+            TEST_OK(device_name, rc=-EINVAL, prom_error,
+                "No device description found in PROM");
+            struct prom_dma_entry *dma_entry =
+                (struct prom_dma_entry *) pentry;
+            device_create(
+                device_class, &pdev->dev, MKDEV(major, minor + i), NULL,
+                "%s.%d.%s", device_name, amc_priv->board, dma_entry->name);
+        }
+        else
+        {
+            printk(KERN_INFO "Unknown entry type found %02x\n", pentry->tag);
+        }
+        i++;
+    }
     return 0;
 
+prom_error:
```

Comment:
Need to keep track of which devices have been created and delete the ones that have been on error.